_posts/2022-02-03-pixel-tests-update.md

```diff
@@ -0,0 +1,136 @@
+rewarding to keep it in good shape, instead of it being so annyoing
+that we would rather ignore it.
+
+## Some numbers
+
+Right now, we have 52 pixel tests in Cockpit itself, 20 in
+Cockpit-machines, and 5 in Cockpit-podman.  Each pixel test has a
+reference image for three layouts, so we have 231 of them, totalling
+about 10 MiB.  The Git repository with all their history is about 22
+MiB, for nine months of pixel testing.  (The main Cockpit repository is
+about 110 MiB, for about nine years of hacking.)
+
+Pixel tests are kind of slow, up to a few seconds per image
+comparison.  They might increase the running time of our tests by a
+couple of minutes.  We will have to measure this more carefully.
+
+## Reproducing pixels
+
+Unsurprisingly, it is non-trivial to make sure that the integration
+tests always perform a given pixel test with exactly the same result.
+We know that we had to battle rendering issues from the start, but it
+took us a bit to figure out the following:
+
+ - Elements need to be scrolled into view before we can test their
+   pixels.  Obvious, but it needed to actually happen before we
+   realized it.
+
+ - There are some animations, and we need to make sure that we don't
+   take a pixel test mid-animation.
+
+ - When the Shell of Cockpit is changed, the content area for pages
+   like Networking might change size.  This is a problem for external
+   projects that don't control the version of the Cockpit shell.  We
+   worked around this by giving the content iframe a fixed size (by
+   making the browser window slightly bigger or smaller if necessary).
+
+ - We would miss failed pixels when a test ran into a known issue. A
+   known issue would cause the test to be skipped, and one would have
+   to look very closely to spot that it also had a failing pixel
+   test. This means that we sometimes forgot to update some reference
+   images, and would later be surprised by how we could have missed
+   them.
+
+ - As we change tests, we would accumulate unused reference images.
+   The bots now force us to clean up properly.
+
+ - Even though the actual DOM rendering is suprisingly deterministic,
+   one or two pixels in a million still come out slightly differently,
+   from time to time.  We have no better idea but to allow a low
+   number of imperfections per pixel test.
+
+With every change, we are more certain that we can get the pixel test
+costs under control.  We might never reach a point where we can stop
+tweaking the pixel test machinery, but the same is true about our CI
+testing machinery in general.
+
+## Why don't we hate them?
+
+I was expecting that adding pixel tests to Cockpit would be a hard
+sell, but we all seem to like them.  That's awesome!  What a team!
+
+I was afraid that pixel tests would be very unreliable because
+browsers would not be nearly pixel perfect enough.  But it turns out
+that they are very good at this, actually.  We had to switch off font
```

Comment:
Even updating browsers to new versions is surprisingly robust.

Reply:
Added.